src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeExtensions.cs

```diff
@@ -19,47 +19,7 @@ internal static class TypeExtensions
         public static MethodInfo GetAnyStaticMethodValidated(this Type type, string name, Type[] types)
         {
             Debug.Assert(types != null);
-            foreach (MethodInfo method in type.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.DeclaredOnly))
-            {
-                if (method.Name == name && method.MatchesArgumentTypes(types))
-                {
-                    return method;
-                }
-            }
-
-            return null;
-        }
-
-        /// <summary>
-        /// Returns true if the method's parameter types are reference assignable from
-        /// the argument types, otherwise false.
-        ///
-        /// An example that can make the method return false is that
-        /// typeof(double).GetMethod("op_Equality", ..., new[] { typeof(double), typeof(int) })
-        /// returns a method with two double parameters, which doesn't match the provided
-        /// argument types.
-        /// </summary>
-        private static bool MatchesArgumentTypes(this MethodInfo mi, Type[] argTypes)
-        {
-            Debug.Assert(mi != null);
-            Debug.Assert(argTypes != null);
-
-            ParameterInfo[] ps = mi.GetParametersCached();
-
-            if (ps.Length != argTypes.Length)
-            {
-                return false;
-            }
-
-            for (int i = 0; i < ps.Length; i++)
-            {
-                if (!TypeUtils.AreReferenceAssignable(ps[i].ParameterType, argTypes[i]))
-                {
-                    return false;
-                }
-            }
-
-            return true;
+            return type.GetMethod(name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.DeclaredOnly, null, types, null);
```

Comment:
Ah ha!

`GetMethod` will match on arithmetic conversions, such as from `int` to `long`. `MatchesArgumentTypes` will not, and should not (such conversions are not handled automatically within Expressions and must be done explicitly).

Therefore we should:
1. Have the main gist of the first commit of this PR.
2. Undo the second commit, so as to keep `MatchesArgumentTypes`, returning null if the result of call to `GetMethod` doesn't pass its requirements.
3. Change the assertion in `MatchesArgumentTypes` that `Debug.Assert(mi != null);` to return null as that can now happen if `GetMethod` returned null.

So the result should be:

```C#
public static MethodInfo GetAnyStaticMethodValidated(this Type type, string name, Type[] types)
{
	Debug.Assert(types != null);
	MethodInfo method = type.GetMethod(name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.DeclaredOnly, null, types, null);
	return method.MatchesArgumentTypes(types) ? method : null;
}

private static bool MatchesArgumentTypes(this MethodInfo mi, Type[] argTypes)
{
	Debug.Assert(argTypes != null);
	if (mi == null)
	{
		return false;
	}
	
	// Rest of this method same as before